provider/mem/mem.go

```diff
@@ -100,37 +101,53 @@ func NewAlerts(ctx context.Context, m types.Marker, intervalGC time.Duration, al
 		logger:    log.With(l, "component", "provider"),
 		callback:  alertCallback,
 	}
-	a.alerts.SetGCCallback(func(alerts []types.Alert) {
-		for _, alert := range alerts {
-			// As we don't persist alerts, we no longer consider them after
-			// they are resolved. Alerts waiting for resolved notifications are
-			// held in memory in aggregation groups redundantly.
-			m.Delete(alert.Fingerprint())
-			a.callback.PostDelete(&alert)
-		}
-
-		a.mtx.Lock()
-		for i, l := range a.listeners {
-			select {
-			case <-l.done:
-				delete(a.listeners, i)
-				close(l.alerts)
-			default:
-				// listener is not closed yet, hence proceed.
-			}
-		}
-		a.mtx.Unlock()
-	})
 
 	if r != nil {
 		a.registerMetrics(r)
 	}
 
-	go a.alerts.Run(ctx, intervalGC)
+	go a.gcLoop(ctx, intervalGC)
```

Comment:
I see we stopped using `a.alerts.SetGCCallback` and `a.alerts.Run`, but I don't understand how that helps fix the race condition? Can we not use the mutex at the start of the callback function?

Reply:
The race condition exists because both the AlertStoreCallback and mem.Alerts.alerts are not protected by the same lock, as I have stated in this PR comment.

> Can we not use the mutex at the start of the callback function?

I do not quite understand this.

Reply:
and the mutex from store.Alerts is actually useless in this context

Reply:
The gccallback occurs after the item is deleted, as shown here: https://github.com/prometheus/alertmanager/blob/c4a763c401742f410580df35a4eedb8daef2a219/store/store.go#L72-L92. The action to delete this item from store.Alerts is not protected by mem.Alerts.mtx. This inconsistency can cause a race condition. We need to ensure that both the callback and the actual delete action are protected by the same lock for the AlertStoreCallback to function properly. Without this, there would be a gap between the delete(store.alerts, xkey) and mem.Alerts.PostDelete actions. This small gap opens up the possibility for a race condition to occur.

Reply:
Ah! I understand the issue! Thanks for the additional explanation, it helped a lot! 🙂

Reply:
I think we can even delete the callback from the `provider/mem/mem.go` as it's not used?

Reply:
Ah it looks like its used in Grafana Mimir.